scalanet/src/io/iohk/scalanet/peergroup/kademlia/KBuckets.scala

```diff
@@ -0,0 +1,109 @@
+package io.iohk.scalanet.peergroup.kademlia
+
+import java.util.concurrent.{ConcurrentHashMap, CopyOnWriteArraySet}
+
+import io.iohk.scalanet.peergroup.kademlia.KBuckets._
+import scodec.bits.BitVector
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+class KBuckets(val baseId: BitVector) {
+
+  private val buckets = new ConcurrentHashMap[Int, KBucket]().asScala
+
+  add(baseId)
+
+  /**
+    * Find the n nodes closest to nodeId in kBuckets.
+    * Return the resulting node records sorted by distance from the nodeId.
+    * The indices into the kBuckets are defined by their distance from referenceNodeId.
+    */
+  def closestNodes(nodeId: BitVector, n: Int): List[BitVector] = {
+    // sketch
+    // the id range of a bucket is 2^i bits,
+    // where i is the common prefix length between the bucket prefix and the base nodeId.
+    // each kbucket covers a distance range [2^i, 2^i+1 - 1]
+    // this means that buckets with larger i cover larger ranges.
+    // so, for eg, bucket 1 covers 4-2-1=1 node whereas bucket 7 covers 256-128-1=127 nodes
+    // in general each bucket covers 2^i+1 - 2^i - 1 = 2^i - 1
+
+    // therefore when finding an arbitrary nodeId
+    // find the prefix
+
+    // replace with routing tree described in kademlia paper...
+    iterator.toList.sorted(new XorOrdering(nodeId)).take(n)
```

Comment:
note that the change on the metric definition affects this part of the code. 
As your metric is not unidirectional, it will be the case that many nodes will be at the same distance. This is not the case with Kademlia's original metric, where all nodes are always at different distance.